applications/FluidDynamicsApplication/custom_strategies/strategies/fs_strategy.h

```diff
@@ -228,164 +155,161 @@ class FSStrategy : public SolvingStrategy<TSparseSpace,TDenseSpace,TLinearSolver
+     */
     virtual void CalculateReactions()
     {
-
-        ModelPart& rModelPart = BaseType::GetModelPart();
-        ProcessInfo& rCurrentProcessInfo = rModelPart.GetProcessInfo();
+        auto &r_model_part = BaseType::GetModelPart();
+        auto &r_process_info = r_model_part.GetProcessInfo();
+        const int n_elems = r_model_part.NumberOfElements();
 
         // Set fractional step index to the momentum equation step
-        int OriginalStep = rCurrentProcessInfo[FRACTIONAL_STEP];
-        rCurrentProcessInfo.SetValue(FRACTIONAL_STEP,1);
-
-#pragma omp parallel
-        {
-            ModelPart::NodeIterator NodesBegin;
-            ModelPart::NodeIterator NodesEnd;
-            OpenMPUtils::PartitionedIterators(rModelPart.Nodes(),NodesBegin,NodesEnd);
-
-            for (ModelPart::NodeIterator itNode = NodesBegin; itNode != NodesEnd; ++itNode)
-            {
-                itNode->FastGetSolutionStepValue(REACTION) = ZeroVector(3);
-            }
-        }
-
-#pragma omp parallel
-        {
-            ModelPart::ElementIterator ElemBegin;
-            ModelPart::ElementIterator ElemEnd;
-            OpenMPUtils::PartitionedIterators(rModelPart.Elements(),ElemBegin,ElemEnd);
-
-            LocalSystemVectorType RHS_Contribution;
-            LocalSystemMatrixType LHS_Contribution;
-
-            for (ModelPart::ElementIterator itElem = ElemBegin; itElem != ElemEnd; ++itElem)
-            {
-
-                //itElem->InitializeNonLinearIteration(rCurrentProcessInfo);
-
-                // Build local system
-                itElem->CalculateLocalSystem(LHS_Contribution, RHS_Contribution, rCurrentProcessInfo);
-
-                Element::GeometryType& rGeom = itElem->GetGeometry();
-                unsigned int NumNodes = rGeom.PointsNumber();
-                unsigned int index = 0;
-
-                for (unsigned int i = 0; i < NumNodes; i++)
-                {
-                    rGeom[i].SetLock();
-                    array_1d<double,3>& rReaction = rGeom[i].FastGetSolutionStepValue(REACTION);
-                    for (unsigned int d = 0; d < mDomainSize; ++d)
-                        rReaction[d] -= RHS_Contribution[index++];
-                    rGeom[i].UnSetLock();
+        const int original_step = r_process_info[FRACTIONAL_STEP];
+        r_process_info.SetValue(FRACTIONAL_STEP, 1);
+
+        // Allocate and initialize values for REACTION calculation
+        LocalSystemVectorType RHS_Contribution;
+        LocalSystemMatrixType LHS_Contribution;
+        VariableUtils().SetHistoricalVariableToZero(REACTION, r_model_part.Nodes());
+
+#pragma omp parallel for private(RHS_Contribution, LHS_Contribution)
+        for (int i_elem = 0; i_elem < n_elems; ++i_elem) {
+            // Build local system
+            auto it_elem = r_model_part.ElementsBegin() + i_elem;
+            it_elem->CalculateLocalSystem(
+                LHS_Contribution,
+                RHS_Contribution,
+                r_process_info);
```

Comment:
I recommend (actually there is no otherway which I had to learn the hard way...) creating a const-ref to the processInfo and pass it here
Otherwise this will screw you up when you start using the new element interfaces

Reply:
BTW isn't this a bit weird, using `CalculateLocalSystem` to compute the reactions?
Wouldn't this be sth for `Calculate`?

Reply:
This might be relevant in #5121

Reply:
> I recommend (actually there is no other way which I had to learn the hard way...) creating a const-ref to the processInfo and pass it here
> Otherwise this will screw you up when you start using the new element interfaces

Done. Thanks!

Reply:
> BTW isn't this a bit weird, using `CalculateLocalSystem` to compute the reactions?
> Wouldn't this be sth for `Calculate`?

The point is that we compute the `REACTION` as minus the residual contribution once the problem has been solved. As the `FractionalStep` element does not implement the RHS-only calculation method, we can only do that by recomputing the local system. I agree this is way inefficient but should be handled in a different PR.

Reply:
> This might be relevant in #5121

mmm... I don't think #5121 influences the `FractionalStepStrategy` as it doesn't require a scheme.